### PR TITLE
Fix SP load issue when authenticated user is from LDAP user store

### PR DESCRIPTION
**Pull request by sadilchamishka**
### Proposed changes in this pull request

$subject

#### Related Issues

- https://github.com/wso2/product-is/issues/4678

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 1]
```diff
@@ -279,6 +281,52 @@ private static boolean isRoleAlreadyApplied(String username, String roleName, Us
...
+            userStoreManager = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getUserStoreManager();
+            if (!(userStoreManager instanceof AbstractUserStoreManager)) {
+                userRoles = userStoreManager.getRoleListOfUser(username);
+            }
+        } catch (UserStoreException e) {
+            throw new IdentityApplicationManagementException("Error while loading the user store manager: ", e);
+        }
+
+        for (ApplicationBasicInfo applicationBasicInfo : applicationBasicInfos) {
+            boolean isAllowedApp = false;
```
Comment (Kanapriya) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:301]: Any reason to have this boolean variable here?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 2]
```diff
@@ -279,6 +281,52 @@ private static boolean isRoleAlreadyApplied(String username, String roleName, Us
...
+            }
+            try {
+                if (userStoreManager instanceof AbstractUserStoreManager &&
+                        ((AbstractUserStoreManager) userStoreManager).isUserInRole(username, applicationRoleName)) {
+                    isAllowedApp = true;
+                    appInfo.add(applicationBasicInfo);
+                } else {
+                    if (Arrays.asList(userRoles).contains(applicationRoleName)) {
+                        isAllowedApp = true;
+                        appInfo.add(applicationBasicInfo);
```
Comment (Kanapriya) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:315]: Shall we merge the above if and else block with the || operator?

Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4394/commits/ed981019cce4dbcdd57b466e3845f918e1555a4f

Merged the if else blocks to a single if-else block

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 3]
```diff
@@ -279,6 +281,50 @@ private static boolean isRoleAlreadyApplied(String username, String roleName, Us
...
+        List<ApplicationBasicInfo> appInfo = new ArrayList<>();
+        UserStoreManager userStoreManager;
+        String[] userRoles = new String[0];
+        try {
+            userStoreManager = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getUserStoreManager();
+            if (!(userStoreManager instanceof AbstractUserStoreManager)) {
+                userRoles = userStoreManager.getRoleListOfUser(username);
+            }
+        } catch (UserStoreException e) {
+            throw new IdentityApplicationManagementException("Error while loading the user store manager: ", e);
```
Comment (Kanapriya) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:297]:
```suggestion
            throw new IdentityApplicationManagementException("Error while retrieving the role list of the user : " +
                    username, e);
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 4]
```diff
@@ -279,6 +281,50 @@ private static boolean isRoleAlreadyApplied(String username, String roleName, Us
...
+            }
+            try {
+                if (userStoreManager instanceof AbstractUserStoreManager &&
+                        ((AbstractUserStoreManager) userStoreManager).isUserInRole(username, applicationRoleName)) {
+                    isAllowedApp = true;
+                    appInfo.add(applicationBasicInfo);
+                } else if (Arrays.asList(userRoles).contains(applicationRoleName)) {
+                    isAllowedApp = true;
+                    appInfo.add(applicationBasicInfo);
+                }
```
Comment (Kanapriya) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:315]: Shall we simplify this logic as follows:
```
if ((userStoreManager instanceof AbstractUserStoreManager &&
                          ((AbstractUserStoreManager) userStoreManager).isUserInRole(username, applicationRoleName))
                      || Arrays.asList(userRoles).contains(applicationRoleName)) {
                  if (log.isDebugEnabled()) {
                      log.debug("Retrieving basic information of authorized application: " +
                              applicationBasicInfo.getApplicationName() + " of the user: " + username);
                  }
                  appInfo.add(applicationBasicInfo);
              }
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 5]
```diff
@@ -279,6 +281,50 @@ private static boolean isRoleAlreadyApplied(String username, String roleName, Us
         return isRoleAlreadyApplied;
     }
 
+    public static List<ApplicationBasicInfo> filterAuthorizedApplicationBasicInfo(
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:292]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4394/commits/5ffd769e055b92ac307c24ca4afe1c2688554b09

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java (hunk 4 above)
Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4394/commits/5ffd769e055b92ac307c24ca4afe1c2688554b09

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 6]
```diff
@@ -279,6 +281,50 @@ private static boolean isRoleAlreadyApplied(String username, String roleName, Us
...
+        } catch (UserStoreException e) {
+            throw new IdentityApplicationManagementException("Error while loading the user store manager: ", e);
+        }
+
+        for (ApplicationBasicInfo applicationBasicInfo : applicationBasicInfos) {
+            boolean isAllowedApp = false;
+            String applicationRoleName = getAppRoleName(applicationBasicInfo.getApplicationName());
+            if (log.isDebugEnabled()) {
+                log.debug("Checking whether user has role : " + applicationRoleName +
+                        " by retrieving role list of user : " + username);
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4394/commits/87c0affd0b58d524002a3d2b26d4f936cf6d7ff7

**Review by Kanapriya — APPROVED** [review]

**Pull request merged**